pennylane/devices/legacy_facade.py

```diff
@@ -231,40 +236,41 @@ def _debugger(self):
     def _debugger(self, new_debugger):
         self._device._debugger = new_debugger
 
-    def preprocess(self, execution_config: ExecutionConfig | None = None):
-        if execution_config is None:
-            execution_config = ExecutionConfig()
+    def preprocess_transforms(
+        self, execution_config: ExecutionConfig | None = None
+    ) -> TransformProgram:
+        program = TransformProgram()
 
-        execution_config = self._setup_execution_config(execution_config)
-        program = qml.transforms.core.TransformProgram()
+        if not execution_config:
+            execution_config = ExecutionConfig()
 
         program.add_transform(legacy_device_batch_transform, device=self._device)
         program.add_transform(legacy_device_expand_fn, device=self._device)
 
         if _requests_adjoint(execution_config):
             _add_adjoint_transforms(program, name=f"{self.name} + adjoint")
 
-        if self._device.capabilities().get("supports_mid_measure", False):
+        if execution_config.mcm_config.mcm_method == "deferred":
+            program.add_transform(
+                defer_measurements,
+                allow_postselect=False,
+            )
```

Comment:
Should we apply defer measurement before `legacy_device_batch_transform` and `legacy_device_expand_fn`?

Reply:
Done